lib/Optimizer/GraphOptimizer/PassManager.cpp

```diff
@@ -0,0 +1,198 @@
+    if (str == otherStr) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Global pass counter used to identify each pass.
+static unsigned globalPassCounter = 0;
+
+} // namespace
+
+bool FunctionPassManager::runPrePass(Function *F,
+                                     const CompilationContext &cctx,
+                                     const FunctionPass &P) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Starting Pass #" << globalPassCounter << ": "
+              << P.getName().str() << " on Function: \""
+              << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGBeforeAllPassesOpt ||
+      listContainsString(dumpDAGBeforePassesOpt, P.getName())) {
+    F->dumpDAG(F->getName().str() + "_PrePass_" + P.getName().str() + "_n" +
+               std::to_string(globalPassCounter) + ".dot");
+  }
+  if (verifyBeforeAllPassesOpt ||
+      listContainsString(verifyBeforePassesOpt, P.getName())) {
+    CHECK(F->verify());
+  }
+  return false;
+}
+
+bool FunctionPassManager::runPostPass(Function *F,
+                                      const CompilationContext &cctx,
+                                      const FunctionPass &P) {
+  if (printPassesOpt) {
+    LOG(INFO) << "Finished Pass #" << globalPassCounter << ": "
+              << P.getName().str() << " on Function: \""
+              << F->getName().str() + "\"\n";
+  }
+  if (dumpDAGAfterAllPassesOpt ||
+      listContainsString(dumpDAGAfterPassesOpt, P.getName())) {
+    F->dumpDAG(F->getName().str() + "_PostPass_" + P.getName().str() + "_n" +
+               std::to_string(globalPassCounter) + ".dot");
+  }
+  if (verifyAfterAllPassesOpt ||
+      listContainsString(verifyAfterPassesOpt, P.getName())) {
+    CHECK(F->verify());
+  }
+  return false;
+}
+
+bool FunctionPassManager::run(Function *F, const CompilationContext &cctx) {
+  bool changed = false;
+
+  for (const FunctionPassConfig &passConfig : pipeline_) {
+    // If we've exceeded the number of passes to run then early exit.
+    if (++globalPassCounter > stopAfterPassNumOpt) {
+      return changed;
+    }
+
+    // Skip some passes if specified by the config that they shouldn't be
+    // executed in this compilation mode.
+    if (!passConfig.isEnabledForCompilationMode(cctx.compMode)) {
+      continue;
+    }
+
+    const FunctionPassID &passID = passConfig.getFunctionPassID();
+
+    // Special case for ConstantFold; skip it if specified.
+    if (passID == FunctionPassID::ConstantFold &&
+        !cctx.optimizationOpts.enableConstantFolding) {
+      continue;
+    }
+
+    // Proceed with performing the pass.
+    auto P = createFunctionPass(passID);
+    changed |= runPrePass(F, cctx, *P);
+
+    switch (passConfig.getConvergenceMode()) {
+    case FunctionPassConfig::ConvergenceMode::OnePass:
+      changed |= P->run(F);
+      break;
+
+    case FunctionPassConfig::ConvergenceMode::UntilFixedPoint:
+      while (P->run(F)) {
+        changed = true;
+        DCE().run(F);
```

Comment:
I like the idea of running `DCE` after each graph-level optimization pass. But I'm not sure if this is the right place to do it. What if we soon decide to use `PassManager` not only for graph-level optimizations, but also for graph-level transformations or IR-level transformations? Maybe it is better to put this DCE call into a `standard` pre- or post-pass action that is executed for all graph FunctionPasses? And all graph function passes derive from some kind of a a GraphFunctionPass? Or we could have a base class for all pass managers and then some derived GraphOptimizationsPassManager, IROptimizationsPassManager, etc. And those derived classes would have these specialized pre/post actions? WDYT?